File: dsh/get-maxkb-token.py
#!/usr/bin/env python3
"""Print a short-lived MaxKB token without persisting its credentials."""

import json
import sys
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


def main() -> int:
    if len(sys.argv) != 3:
        print(f"Usage: {Path(sys.argv[0]).name} LAUNCHER_PS1 MAXKB_URL", file=sys.stderr)
        return 2
    launcher = Path(sys.argv[1])
    base_url = sys.argv[2].rstrip("/")
    try:
        account_file = launcher.parent.parent / "01-环境与账号" / "maxkb-account.json"
        account = json.loads(account_file.read_text(encoding="utf-8-sig"))
        request = Request(
            f"{base_url}/admin/api/user/login",
            data=json.dumps({"username": account["username"], "password": account["password"]}).encode(),
            headers={"Content-Type": "application/json"}, method="POST",
        )
        with urlopen(request, timeout=10) as response:
            token = json.load(response).get("data", {}).get("token")
        if not token:
            raise ValueError("MaxKB login did not return a token")
    except (KeyError, OSError, ValueError, HTTPError, URLError, json.JSONDecodeError) as error:
        print(f"Unable to obtain MaxKB token: {error}", file=sys.stderr)
        return 1
    print(token)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: dsh/get-maxkb-token.mjs
#!/usr/bin/env node
import { readFile } from 'node:fs/promises'

const [accountPath, baseUrl = process.env.MAXKB_BASE_URL || 'http://127.0.0.1:8080'] = process.argv.slice(2)
if (!accountPath) throw new Error('Usage: get-maxkb-token.mjs ACCOUNT_JSON [MAXKB_URL]')
const account = JSON.parse((await readFile(accountPath, 'utf8')).replace(/^\uFEFF/, ''))
const response = await fetch(`${baseUrl.replace(/\/+$/, '')}/admin/api/user/login`, {
  method: 'POST', headers: { 'content-type': 'application/json' },
  body: JSON.stringify({ username: account.username, password: account.password }),
})
const body = await response.json()
const token = body?.data?.token
if (!response.ok || !token) throw new Error(body?.message || 'MaxKB login did not return a token')
process.stdout.write(String(token))
